linkerd/router/src/lib.rs

```diff
@@ -0,0 +1,162 @@
+use ahash::AHashMap;
+use futures::{future, prelude::*};
+use linkerd_error::Error;
+use linkerd_stack::{layer, NewService, Oneshot, Param, Service, ServiceExt};
+use std::{
+    fmt::Debug,
+    hash::Hash,
+    marker::PhantomData,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tracing::{debug, error};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct RouteKeys<K>(Arc<[K]>)
+where
+    K: Eq + Hash + Clone;
+
+pub trait SelectRoute<Req> {
+    type Key: Eq + Hash + Clone + Debug + Send + Sync + 'static;
+    type Error: std::error::Error + Send + Sync + 'static;
+
+    /// Given a a request, returns the key matching this request.
+    ///
+    /// If no route matches the request, this method returns an error.
+    fn select<'r>(&self, req: &'r Req) -> Result<&Self::Key, Self::Error>;
+}
+
+/// A [`NewService`] that extracts [`RouteKeys`] from the stack target to build a
+/// set of per-route inner services. These services are built first by building
+/// a backend stack for the target and then applying the `L`-typed router layer
+/// to the backend stack.
+#[derive(Clone, Debug)]
+pub struct NewRoute<K, L, N> {
+    route_layer: L,
+    inner: N,
+    _marker: PhantomData<fn(K)>,
+}
+
+/// The [`Service`] constructed by [`NewRoute`].
+///
+/// Each request is matched against the route table and routed to the
+/// appropriate
+#[derive(Clone, Debug)]
+pub struct Route<T, K, S> {
+    params: T,
+    routes: Arc<AHashMap<K, S>>,
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("unknown route: {0:?}")]
+pub struct UnknownRoute<K: std::fmt::Debug>(K);
+
+// === impl NewRoute ===
+
+impl<K, L: Clone, N> NewRoute<K, L, N> {
+    pub fn new(route_layer: L, inner: N) -> Self {
+        Self {
+            inner,
+            route_layer,
+            _marker: PhantomData,
+        }
+    }
+
+    pub fn layer(route_layer: L) -> impl layer::Layer<N, Service = Self> + Clone {
+        layer::mk(move |inner| Self::new(route_layer.clone(), inner))
+    }
+}
+
+impl<T, K, L, N, S> NewService<T> for NewRoute<K, L, N>
+where
+    K: Eq + Hash + Clone + Debug + Send + Sync + 'static,
+    T: Param<RouteKeys<K>> + Clone,
+    N: NewService<T>,
+    L: layer::Layer<N::Service>,
+    L::Service: NewService<K, Service = S>,
+{
+    type Service = Route<T, K, S>;
+
+    fn new_service(&self, target: T) -> Self::Service {
+        let inner = self.inner.new_service(target.clone());
+        let new_route = self.route_layer.layer(inner);
+
+        let RouteKeys(keys) = target.param();
+        let routes = keys
+            .iter()
+            .map(|key| (key.clone(), new_route.new_service(key.clone())))
+            .collect();
```

Comment:
We could potentially eliminate this and instead lazily cache inner services as they're referenced.